src/backend/gporca/libgpopt/src/xforms/CXformFactory.cpp

```diff
@@ -405,18 +404,18 @@ CXformFactory::Init()
 void
 CXformFactory::Shutdown()
 {
-	CXformFactory *pxff = CXformFactory::Pxff();
+	CXformFactory *pxff = m_pxff;
 
 	GPOS_ASSERT(NULL != pxff && "Xform factory has not been initialized");
 
 	CMemoryPool *mp = pxff->m_mp;
```

Comment:
Does the order of the commands, executed inside Shutdown method (not only in this class, but at whole classes changed in this patch) is significant? In this case m_pxff would be set to NULL, and on release build the excepton maybe raised, but other classes has a different order (for ex. [here](https://github.com/arenadata/gpdb/blob/5681fe94a83080e60d8b361bd9b6e8d6bbe90969/src/backend/gporca/libgpos/src/common/CDebugCounter.cpp#L79), or the CMessageRepository), and the static variable may be not cleaned up. For me it's a bit confusing, may be we shouldn't do the same code at the shutdown methods, but we should use the single order of cleanup?

Reply:
> Does the order of the commands, executed inside Shutdown method (not only in this class, but at whole classes changed in this patch) is significant?

Yes, because some destructors have corresponding asserts.

Reply:
I Understood, other classes which doesn't have an assertion like `GPOS_ASSERT(global_instalnce == NULL)` inside descturcor may not nulls the variable as a first command, so them may call the possble  method (which may throw exception) before nulling, and seems it's a valid case: Shutdown method may throw an exception before it nulls the global_instance, but the desctructor won't fail on assertion (on debug build) for example [CFSImulator::Shutdown](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/src/error/CFSimulator.cpp#L158-L166) may throw an exception on GPOS_DEBUG, but it's [destructor](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/gporca/libgpos/include/gpos/error/CFSimulator.h#L165-L167) is empty